### PR TITLE
EKA-136 | Handling error scenarios, during processing of data.

### PR DESCRIPTION
**Pull request by angshu**

### CODE REVIEW

File: src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java [hunk 1]
```diff
@@ -119,16 +121,30 @@ public DataFlowRepository(PgPool pgPool) {
...
+                dbClient.preparedQuery(
+                        UPDATE_HEALTH_DATA_AVAILABILITY,
+                        Tuple.of(status.toString(), allErrors, transactionId, dataPartNumber),
+                        handler -> {
+                            if (handler.failed())
+                                monoSink.error(new Exception("Failed to update health data availability"));
+                            else
+                                monoSink.success();
                         })
         );
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/DataFlowRepository.java]:
```suggestion
                                monoSink.success();
                        }));
```

File: src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java [hunk 2]
```diff
@@ -27,11 +28,11 @@
 
     private Flux<DataEntry> getDataEntries(String transactionId, String hipId, String hipName) {
         return healthInformationRepository.getHealthInformation(transactionId)
-                .map(data -> DataEntry.builder()
+                .map(healthInfo -> DataEntry.builder()
                         .hipId(hipId)
                         .hipName(hipName)
-                        //.status(data != null ? Status.COMPLETED : Status.REQUESTED)
-                        .data(data)
+                        .status(EntryStatus.valueOf((String) healthInfo.get("status")))
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java]: What happens if `healthInfo.get("status")` is not one of the possible values?

File: src/main/java/in/org/projecteka/hiu/dataflow/model/DataEntry.java [hunk 3]
```diff
@@ -12,6 +13,6 @@
 public class DataEntry {
     private String hipId;
     private String hipName;
-    private Status status;
+    private EntryStatus status;
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataflow/model/DataEntry.java:16]: For me status made more sense than having redundant prefix Entry, as it was inside DataEntry object already, it was supposed to represent DataEntry status only.

File: src/main/java/in/org/projecteka/hiu/dataprocessor/DataAvailabilityListener.java [hunk 4]
```diff
@@ -46,11 +46,9 @@ public void subscribe()  {
             DataAvailableMessage dataAvailableMessage = deserializeMessage(message);
             logger.info(String.format("Received notification of data availability for transaction id : %s", dataAvailableMessage.getTransactionId()));
             logger.info(String.format("Processing data from file : %s", dataAvailableMessage.getPathToFile()));
-            try {
-                new HealthDataProcessor(healthDataRepository, dataFlowRepository, new Decryptor()).process(dataAvailableMessage);
-            } catch (Exception e) {
-                e.printStackTrace();
-            }
+            HealthDataProcessor healthDataProcessor = new HealthDataProcessor(healthDataRepository, dataFlowRepository, new Decryptor());
+            healthDataProcessor.registerHITypeResourceHandler(new DiagnosticReportResourceProcessor());
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataprocessor/DataAvailabilityListener.java]: Why can't this also be constructor parameter?

File: src/main/java/in/org/projecteka/hiu/dataprocessor/HealthDataProcessor.java [hunk 5]
```diff
@@ -72,13 +71,36 @@ private void processEntries(DataContext context) {
...
             //TODO: else part. download the content from entry.getLink().getHref(), and essentially call processEntryContent()
         });
+
+        if (!dataErrors.isEmpty()) {
+            String allErrors = "[ERROR]".concat(String.join("[ERROR]", dataErrors));
+            updateDataProcessStatus(context, allErrors, HealthInfoStatus.ERRORED);
+        } else {
+            updateDataProcessStatus(context, "", HealthInfoStatus.SUCCEEDED);
+        }
+
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/dataprocessor/HealthDataProcessor.java]:
```suggestion
```

File: src/main/java/in/org/projecteka/hiu/dataprocessor/DataAvailabilityListener.java (hunk 4 above)
Comment (angshu) [reply src/main/java/in/org/projecteka/hiu/dataprocessor/DataAvailabilityListener.java]: There would be potentially many such processors. or we take a list<Processor> as constructor args

File: src/main/java/in/org/projecteka/hiu/dataflow/model/DataEntry.java (hunk 3 above)
Comment (angshu) [reply src/main/java/in/org/projecteka/hiu/dataflow/model/DataEntry.java:16]: there are 2 different status fields .. we wanted to keep the distinction.

File: src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java (hunk 2 above)
Comment (angshu) [reply src/main/java/in/org/projecteka/hiu/dataflow/HealthInfoManager.java]: there ideally would never be a scenario of the status field not being present or being invalid. but ok, can be better checked

**Review by ganesanarun — APPROVED** [review]

**Pull request merged**